### ISSUE TITLE
Add support for Python UDFs in distributed queries

### ISSUE THREAD

**Issue by andygrove** [apache/arrow-ballista#173]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I would like to be able to execute queries containing custom Python code. This is already supported in DataFusion but we need to add the serde aspect in Ballista.

**Describe the solution you'd like**
It looks like we can serialize Python functions with https://docs.rs/pyo3/0.7.0/pyo3/marshal/index.html

We then need to store the serialied Python functions in protobuf and deserialize them in the executors.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by avantgardnerio** [comment]
Fancy!

**Comment by adriangb** [comment]
Would the plan be to focus on Python UDFs that don't have dependencies? I don't think I've ever seen a good story for taking Python code from one place and executing it in another if there are 3rd party dependencies involved. Would you folks consider some sort of HTTP UDF as well? [BigQuery supports a similar feature](https://cloud.google.com/bigquery/docs/reference/standard-sql/remote-functions). It could probably be done better here, e.g. by transmitting over the wire in a more efficient format and providing tooling to deserialize on the other end (maybe DataFusion?).

**Comment by andygrove** [comment]
Unfortunately, I do not have the Python skills to answer the question. I think we can learn from how Dask does this? I wonder if @jdye64 can offer any high-level guidance.

**Comment by andygrove** [comment]
Another approach here is to implement a custom executor process that wraps Ballista and takes care of registering custom Python code as UDFs so that we don't have to worry about trying to send code over the network.

**Comment by jdye64** [comment]
Yes, @adriangb is right. Much pain comes in trying to serialize and execute python code on remote nodes that have dependencies. This has been the case since even for Hive UDFs back years ago.

The python ecosystem as a whole is one that relies heavily on existing dependencies. Therefore I think if we can come up with a straightforward method for ensuring all of the executors have a valid virtual environment with all the dependencies required by the UDF installed we should be good. This is the approach we take in some parts of Dask for example.

So maybe as part of the Python UDF registration we require a "list" of dependencies that are required by the UDF. When the executor server starts up it could create that virtual env, through pip or conda or whatever is chosen, and installed those dependencies. Think of it like a executor server bootstrapping process. Then when any sql queries are submitted the UDF can be serialized and sent to the executor, once there the UDF can be executed in that virtual environment.

Couple of thoughts
- Maybe that information about Python dependencies could live in the "catalog" description space of flight_sql in ballista?
- I think being able to run Python UDFs is a must, almost not even worth having Python UDF support if dependencies can't be used. This is just my opinion and not a fact.
- I can remember old versions of Hive required the user to manually SSH to each node and manually install those python dependencies. It was the quickest route I ever discovered to making enemies in the dev ops teams =) . I think this path is a non starter.

**Comment by adriangb** [comment]
> I think being able to run Python UDFs is a must, almost not even worth having Python UDF support if dependencies can't be used. This is just my opinion and not a fact.

Agreed that if you want to allow custom Python code to run you need to allow 3rd party dependencies. But 3rd party dependencies are a whole big can of worms in Python, to the point where I would avoid opening it if you can. Hence the suggestion for HTTP UDFs which also have other benefits/use cases.

I think a key aspect of this is to allow users to stick to the workflows they know instead of having to build a new one. For example as a data engineer/backend dev I manage a large project with multiple deployable artifacts that get bundled into containers and deployed on k8s. I have the knowledge and infrastructure in place to handle all of the complexities involved in this (e.g. locking dependencies across deployable alá Cargo workspaces). Any sort of new dependency management paradigm that does not fit in with this is extra work and possibly a source of bugs. That includes the SSH into a node model and Airflow's same dependency everywhere model 🙃. I think a good model would be a sort of "UDF executor web framework" that does some hand-holding but ultimately leaves packaging and dependencies up to users:

```python
from ballista import UDFExecutorApp, udf
from pyarrow import RecordBatchReader

@udf.aggregator(name="override_the_name")
def some_aggregation_function(reader: RecordBatchReader) -> Iterable[RecordBatch]:
    ...

def main() -> None:
    app = UDFExecutorApp([some_aggregation_function])
    app.serve(scheduler_host=...)
```

I'm just making something up here but the point is to keep a relatively familiar Flask-style API but abstract away the fiddly bits.

This would register itself with the scheduler as being able to execute `"override_the_name"`.

**Comment by jdye64** [comment]
I have no hands-on experience with the HTTP UDFs model but am intrigued by the approach. Thank you for laying out your thoughts. Had some thoughts and questions.

#### Things I really like:
- Clearly a remote invocation approach makes managing 3rd party dependencies a breeze
- Clean and repeatable model for creating UDFs. I think this approach would be fairly intuitive for any Python user.

#### Concerns/Questions (mostly because of my lack of experience here)
- Where would the HTTP server be hosted? Scheduler? Single Executor process? Multiple Executor processes? An entirely new process?
- Wouldn't it make more sense to have that communication channel be Arrow flight instead of HTTP?
- Seems like this would introduce a large performance hit having to make an external (data movement) remote invocation for each RecordBatch since the data would need to be moved to the HTTP UDF service running on another host.

**Comment by adriangb** [comment]
> Where would the HTTP server be hosted? Scheduler? Single Executor process? Multiple Executor processes? An entirely new process?

I'd think it'd be very similar to an executor process (a UDF executor process?). But I'm not super familiar with how executors run now, how they scale, etc.

> Wouldn't it make more sense to have that communication channel be Arrow flight instead of HTTP?

With this example above I'd say it can be Arrow flight or any other communication protocol, I'm not constraining it to HTTP.
It would be really cool though if as a user I could deploy one thing that can talk over Arrow Flight for use as a UDF but also serve an HTTP API.

> Seems like this would introduce a large performance hit having to make an external (data movement) remote invocation for each RecordBatch since the data would need to be moved to the HTTP UDF service running on another host.

Yeah this I don't know about. Would the data be currently residing in the Executor? If so the only way to do everything in memory would be to run these UDFs on the Executor itself. Which gets into the complication of dependency management. Could users wrap the executor itself? Maybe when `UDFExecutorApp` registers itself with the scheduler it _becomes_ an executor which execute regular queries but also the UDFs that it is running locally? It might get confusing if different executors have different UDFs available...

**Comment by eliyahweinberg** [comment]
In Apache Spark, we utilize Python PEX executables that are built with all necessary dependencies. When an executor is launched, it downloads the PEX file during initialization, which is then used to execute Python UDF code.

**Comment by milenkovicm** [comment]
Latest ballista release can be extended to support python udf(s). A small poc to demonstrate https://github.com/milenkovicm/ballista_python

**Comment by matthewmturner** [comment]
This is really cool.  I havent followed the whole conversation but is the `PythonFunctionFactory` available for other apps to use? I'm quite keen to get it added to dft.

**Comment by milenkovicm** [comment]
`PythonFunctionFactory` is part of showcase repo, it  is very basic I haven't put much effort into it. Maybe we could put some more effort into it.

On the similar note, how can we add ballista support into dft?

**Comment by matthewmturner** [comment]
Maybe `PythonFunctionFactory` could be added to `datafusion-python`?

Before my refactor last summer dft used to support ballista and ive had in mind adding it back - but ive had other priorities.  If it's something i know would be used by others im not opposed to pushing it up the list.  Generally speaking i think it would be added in a similar fashion to how FlightSQL support was added - which is basically adding a new feature gated context type to the apps `AppExecution` and implementing the relevant methods for executing & benchmarking queries.

**Comment by matthewmturner** [comment]
And to add - i had in mind starting with just making dft a ballista client.  Id have to give more thought to what it would look like if it were to be used as part of the ballista cluster.

**Comment by milenkovicm** [comment]
> Maybe `PythonFunctionFactory` could be added to `datafusion-python`?

we should run that idea in datafusion python discord

> Before my refactor last summer dft used to support ballista and ive had in mind adding it back - but ive had other priorities. If it's something i know would be used by others im not opposed to pushing it up the list. Generally speaking i think it would be added in a similar fashion to how FlightSQL support was added - which is basically adding a new feature gated context type to the apps `AppExecution` and implementing the relevant methods for executing & benchmarking queries.

not sure if you're up to data, ballista this days uses datafusion SessionContext so it should work no problem

**Comment by matthewmturner** [comment]
> we should run that idea in datafusion python discord

Sure will raise it there.

Im also okay to incubate `PythonFunctionFactory` in dft to start, im doing that for [WASM UDFs](https://github.com/datafusion-contrib/datafusion-dft/tree/main/crates/datafusion-udfs-wasm) already.
 
> not sure if you're up to data, ballista this days uses datafusion SessionContext so it should work no problem

I do recall seeing that - should make things easier and potentially plug and play with existing local context.  Will have to check it out.

I'm currently working on some auth capabilities which i expect to finish up in the coming week.  If youre open to helping test the ballista implementation once I finish that I could work on it next.

**Comment by milenkovicm** [comment]
cc-ing: @timsaucer

**Comment by amimardev** [comment]
Will this support zero copy python udfs ?
And is this possible to do in ballista ?
[Lakesail blog] (https://lakesail.com/blog/sail-and-python-udfs/)

**Comment by milenkovicm** [comment]
> Will this support zero copy python udfs ? And is this possible to do in ballista ? [Lakesail blog] (https://lakesail.com/blog/sail-and-python-udfs/)

not in valilla ballista, but it can be extended to support it. example how to do it can be found at https://github.com/milenkovicm/ballista_python 

yes its zero data copy using arrow-c interface

**Comment by amimardev** [comment]
> > Will this support zero copy python udfs ? And is this possible to do in ballista ? [Lakesail blog] (https://lakesail.com/blog/sail-and-python-udfs/)
> 
> not in valilla ballista, but it can be extended to support it. example how to do it can be found at https://github.com/milenkovicm/ballista_python
> 
> yes its zero data copy using arrow-c interface

Also, lakesail is saying that it has 94% les sresouece usage then spark,
What about ballista ?
